pkcs5/src/pbes2.rs

```diff
@@ -0,0 +1,223 @@
+    /// Password-Based Key Derivation Function 2 (PBKDF2).
+    Pbkdf2(Pbkdf2Params<'a>),
+
+    /// Other key derivation functions.
+    Other(AlgorithmIdentifier<'a>),
+}
+
+impl<'a> TryFrom<AlgorithmIdentifier<'a>> for Kdf<'a> {
+    type Error = der::Error;
+
+    fn try_from(alg: AlgorithmIdentifier<'a>) -> der::Result<Self> {
+        match alg.oid {
+            PBKDF2_OID => alg
+                .parameters_any()
+                .and_then(TryFrom::try_from)
+                .map(Self::Pbkdf2),
+            _ => Ok(Self::Other(alg)),
+        }
+    }
+}
+
+impl<'a> Kdf<'a> {
+    /// Get [`Pbkdf2Params`] if it is the selected algorithm.
+    pub fn pbkdf2(self) -> Option<Pbkdf2Params<'a>> {
+        match self {
+            Self::Pbkdf2(params) => Some(params),
+            _ => None,
+        }
+    }
+}
+
+/// Password-Based Encryption Scheme 2 parameters as defined in [RFC 8018 Appendix A.2].
+///
+/// ```text
+/// PBKDF2-params ::= SEQUENCE {
+///     salt CHOICE {
+///         specified OCTET STRING,
+///         otherSource AlgorithmIdentifier {{PBKDF2-SaltSources}}
+///     },
+///     iterationCount INTEGER (1..MAX),
+///     keyLength INTEGER (1..MAX) OPTIONAL,
+///     prf AlgorithmIdentifier {{PBKDF2-PRFs}} DEFAULT
+///     algid-hmacWithSHA1 }
+/// ```
+///
+/// [RFC 8018 Appendix A.2]: https://tools.ietf.org/html/rfc8018#appendix-A.2
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Pbkdf2Params<'a> {
+    /// PBKDF2 salt
+    // TODO(tarcieri): support `CHOICE` with `otherSource`
+    pub salt: &'a [u8],
+
+    /// PBKDF2 iteration count
+    pub iteration_count: u16,
+
+    /// PBKDF2 output length
+    // TODO(tarcieri): support this OPTIONAL field
+    // Blocked on: https://github.com/RustCrypto/utils/issues/271
+    pub key_length: Option<u16>,
+
+    /// Pseudo-random function to use with PBKDF2
+    pub prf: Pbkdf2Prf,
+}
+
+impl<'a> TryFrom<der::Any<'a>> for Pbkdf2Params<'a> {
+    type Error = der::Error;
+
+    fn try_from(any: der::Any<'a>) -> der::Result<Self> {
+        any.sequence(|params| {
+            // TODO(tarcieri): support salt `CHOICE` w\ `AlgorithmIdentifier`
+            let salt = params.octet_string()?;
+            let iteration_count = params.decode()?;
+
+            // TODO(tarcieri): support OPTIONAL key length field
+            // Blocked on: https://github.com/RustCrypto/utils/issues/271
+            let key_length = None;
+            let prf = AlgorithmIdentifier::decode(params)?;
+
+            Ok(Self {
+                salt: salt.as_bytes(),
+                iteration_count,
+                key_length,
+                prf: prf.try_into()?,
+            })
+        })
+    }
+}
+
+/// Pseudo-random function used by PBKDF2.
+// TODO(tarcieri): support default `algid-hmacWithSHA1`?
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum Pbkdf2Prf {
+    /// HMAC with SHA-256
+    HmacWithSha256,
+
+    /// Other PRFs
+    Other(ObjectIdentifier),
```

Comment:
Several PBES2 enums include these catch-all `Other` variants, as this PR only implements a single ciphersuite (PBKDF2-SHA256 w\ AES-256-CBC).

It would be good exhaustively implement all of the algorithms described in RFC 8018, however even then the RFC says "Applications may employ other schemes as well" so it seems these should probably remain.

Reply:
What is a hypothetical use-case for those `Other` variants instead of simply returning an "unknown algorithm" error immediately? It looks a bit strange to have `#[non_exhaustive]` with `Other` variant.

Reply:
Per the RFC, it'd be an application-specific use case. Note that this one in particular is fairly deep within the structure.

Simply returning an error in these cases is an alternative, although that said, the `der` crate doesn't presently have a good way to propagate an "unknown OID" error (which is probably something that should be added, and can be done in a non-breaking way).

Reply:
Went ahead and removed the `Other` variants and now return `der::ErrorKind::OidInvalid` (added in #273) which wraps the invalid OID value and returns it back to the caller.

The remaining `enum`s in the `pbes2` module are all marked `#[non_exhaustive]` so we can incrementally add additional algorithm support (although especially for PBES2, I'd like to also add test vectors for each algorithm we add)